apm-agent-plugins/apm-okhttp-plugin/src/main/java/co/elastic/apm/agent/okhttp/OkHttp3ClientAsyncInstrumentation.java

```diff
@@ -0,0 +1,165 @@
+
+    @VisibleForAdvice
+    public static final Logger logger = LoggerFactory.getLogger(OkHttp3ClientAsyncInstrumentation.class);
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return OkHttpClient3ExecuteAdvice.class;
+    }
+
+    @Nullable
+    @VisibleForAdvice
+    public static HelperClassManager<WrapperCreator<Callback>> callbackWrapperCreator;
+
+
+    @Override
+    public void init(ElasticApmTracer tracer) {
+        callbackWrapperCreator = HelperClassManager.ForAnyClassLoader.of(tracer,
+            OkHttp3ClientAsyncInstrumentation.class.getName() + "$CallbackWrapperCreator",
+            OkHttp3ClientAsyncInstrumentation.class.getName() + "$CallbackWrapperCreator$CallbackWrapper");
+    }
+
+    @VisibleForAdvice
+    public static class OkHttpClient3ExecuteAdvice {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void onBeforeEnqueue(@Advice.Origin Class<? extends Call> clazz,
+                                            @Advice.FieldValue(value = "originalRequest", typing = Assigner.Typing.DYNAMIC, readOnly = false) @Nullable okhttp3.Request originalRequest,
+                                            @Advice.Argument(value = 0, readOnly = false) @Nullable Callback callback,
+                                            @Advice.Local("span") Span span) {
+            if (tracer == null || tracer.getActive() == null || callbackWrapperCreator == null) {
+                return;
+            }
+
+            final WrapperCreator<Callback> wrapperCreator = callbackWrapperCreator.getForClassLoaderOfClass(clazz);
+            if (originalRequest == null || callback == null || wrapperCreator == null) {
+                return;
+            }
+
+            final TraceContextHolder<?> parent = tracer.getActive();
+
+            okhttp3.Request request = originalRequest;
+            span = HttpClientHelper.startHttpClientSpan(parent, request.method(), request.url().toString(), request.url().host());
+            if (span != null) {
+                span.activate().markLifecycleManagingThreadSwitchExpected();
+                originalRequest = originalRequest.newBuilder().addHeader(TraceContext.TRACE_PARENT_HEADER, span.getTraceContext().getOutgoingTraceParentHeader().toString()).build();
+                callback = wrapperCreator.wrap(callback, span);
+            }
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class)
+        private static void onAfterEnqueue(@Advice.Local("span") @Nullable Span span) {
+            if (span != null) {
+                span.deactivate();
+            }
+        }
+    }
+
+    public static class CallbackWrapperCreator implements WrapperCreator<Callback> {
+
+        @Override
+        public Callback wrap(final Callback delegate, Span span) {
+            return new CallbackWrapper(span, delegate);
+        }
+
+        private static class CallbackWrapper implements Callback {
+            private final Span span;
```

Comment:
I wrote this comment in the initial review and it got lost somehow, noticed just now. 

Like we did with the ES client, we need to enforce proper visibility. Make this `span` field volatile and switch the order of assignments in the constructor (so that it is written to after `delegate` is assigned on the initiating thread). In both `onFailure` and `onResponse` the `span` is read first, so this should suffice. Otherwise, you can make both `span` and `delegate` volatile.

Reply:
We don't need to in this case, because the fields are `final` 🙂

Reply:
Right 🤦‍♂️ , in my head we were recycling them 😄 . 
I don't have support for async discussions yet, hence the visibility problem 🤓